Add ability to limit fields added during Mapper#merge

felixbarny opened the pull request:
This is extracting a piece of independent functionality from https://github.com/elastic/elasticsearch/pull/96235

This is the result of this discussion: https://github.com/elastic/elasticsearch/pull/96235#discussion_r1394694066

The mapping merge executed on the master node is the only authoritative place where we have a consistent view of the cluster state as it's guaranteed to be executed on a single thread on the master node. Therefore, there can be no race conditions and we exactly know how many fields are left before hitting the field limit.

Therefore, this is the best place to add as many fields as possible without going over the limit if `index.mapping.total_fields.ignore_dynamic_beyond_limit` is set to `true` in the index settings.

This change makes it possible to limit the number of fields added during mapping merger. However, that feature isn't actually used outside of tests.

In https://github.com/elastic/elasticsearch/pull/96235, the feature to limit fields during the merge process is used to guarantee that the cluster state update never fails due to dynamic mapping updates that exceed the field limit in case `ignore_dynamic_beyond_limit` is enabled.

This works because after each dynamic mapping update, the indexing request gets retried (there's no change to that logic in this or related PRs) so that the document parsing takes the new mapping into account.

This only applies to dynamic mapping updates aka auto-updates. When explicitly trying to update the mapping with too many fields, the put mapping operation will still fail, even if `ignore_dynamic_beyond_limit` is enabled.


Depends on:
- [x] https://github.com/elastic/elasticsearch/pull/104512
- [x] https://github.com/elastic/elasticsearch/pull/104578

felixbarny committed:
Add ability to limit fields added during Mapper#merge

This is extracting a piece of independent functionality from https://github.com/elastic/elasticsearch/pull/96235

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

felixbarny commented on server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:
```diff
@@ -536,17 +546,25 @@ private static Map<String, Mapper> buildMergedMappers(
+            MapperMergeContext objectMergeContext
         ) {
-            Map<String, Mapper> mergedMappers = null;
+            // need to preserve order to make it deterministic which fields are ignored
+            Map<String, Mapper> mergedMappers = new LinkedHashMap<>(existing.mappers);
```
While I think that making it more deterministic which fields get added vs ignored makes sense, I don't think it's a huge issue if it isn't. So if you think this would cause more issues than it solves (such as performance and memory impact), I can just revert this to a regular `HashMap` and use `Map.copyOf` instead of `Collections.unmodifiableMap`. I think the biggest benefit is that it does make the testing a bit easier if you know deterministically, which fields you can expect to be in the mapping vs which ones get ignored.

felixbarny replied:
I've removed the new order-preserving semantics and adjusted the tests so that they don't assume ordering.

felixbarny committed:
Add unit tests

felixbarny committed:
Don't preserve order in mappers

felixbarny committed:
Apply spotless suggestions

elasticmachine committed:
Merge branch 'main' into mapping-merge-limit-fields

felixbarny committed:
Comment out new tests in ObjectMapperMergeTests to check if they're the cause of the segfault

felixbarny committed:
Revert "Comment out new tests in ObjectMapperMergeTests to check if they're the cause of the segfault"

This reverts commit 9c49a87e450ad88bc058204b167fa640ed568503.

felixbarny committed:
Merge remote-tracking branch 'origin/main' into mapping-merge-limit-fields

felixbarny committed:
Merge remote-tracking branch 'origin/main' into mapping-merge-limit-fields

felixbarny committed:
Consider runtime fields for root mapper size

felixbarny committed:
Assert mapper size equal to builder size

felixbarny committed:
Fix SemanticTextFieldMapperTests

felixbarny committed:
Fix multi-field mapper builder size calculation

felixbarny committed:
Apply spotless suggestions

felixbarny committed:
Fixes and polishing

felixbarny committed:
Fix SemanticTextFieldMapperTests

felixbarny committed:
Merge remote-tracking branch 'origin/main' into mapping-merge-limit-fields

felixbarny committed:
Revert "Apply spotless suggestions"

This reverts commit 2e0f00ca07f1723b8b45b6c32d337843f85b5a4d.

felixbarny committed:
Revert "Fix multi-field mapper builder size calculation"

This reverts commit fd524dc085f2f40cb25dbbcfd25e8db59159c51a.

felixbarny committed:
Revert "Assert mapper size equal to builder size"

This reverts commit befd7f6802ed2ec2eb1e05d16a9f64c0c46b971e.

felixbarny committed:
Merge remote-tracking branch 'origin/main' into mapping-merge-limit-fields

felixbarny committed:
Remove Mapper.Builder#mapperSize

Because of https://github.com/elastic/elasticsearch/pull/103865,
DocumentParserContext#addDynamicMapper receives a Mapper, not a Mapper.Builder again.
Therefore, we don't need a mapperSize method for the builder.

This simplifies things a lot.

felixbarny committed:
Remove unused import

javanna reviewed (commented):
I started a round of review. I haven't completed it yet, but I have two comments so far that I want to send out while I finish my review.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/MapperMergeContext.java:
```diff
@@ -0,0 +1,84 @@
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class MapperMergeContext {
```
While introducing this new context may be a good idea (it is weird to use MapperBuilderContext when merging to some degree), I find that it requires a lot of mechanical changes that end up being a distraction when reviewing the core changes in this PR. Is it an option to consider making that as a follow-up perhaps? Do you consider using a separate context a requirement for this PR to go in? The point is that the builder context has stuff that is not needed to merge, and you are adding stuff that is only needed for merging? If we think it's needed, we could also make that change as a pre-requisite, it should be easier to review on its own.

felixbarny replied:
> Is it an option to consider making that as a follow-up perhaps?

I think that may lead to even more mechanical changes as we'd need to add `maxFieldsToAddDuringMerge` to `MapperBuilderContext` and then move it to `MapperBuilderContext` again.

But I could probably introduce the `MapperMergeContext` in a separate PR, as a prerequisite of this PR. On the other hand, this PR isn't exactly huge, even if it contains some mechanical changes.

But if you prefer so, I'll go ahead and do it. Let me know.

javanna replied:
The PR isn't huge, I agree, but not small either. The problem is when we mix mechanical changes with core changes that are delicate. I am afraid of missing something, especially as we have many merge methods and it's difficult to track which are used where. I'd be happy to review adding this MapperMergeContext separately if that's not too much work on your end.

felixbarny replied:
np, I'm working on it!

felixbarny replied:
Here's the PR: https://github.com/elastic/elasticsearch/pull/104512
Once that is merged, I'll update this PR which should drastically reduce the number of mechanical changes, exposing just the changes to the mapping merge process.

javanna replied:
++ thanks

felixbarny commented on server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:
```diff
@@ -558,12 +558,25 @@ public Mapping parseMapping(String mappingType, Map<String, Object> mappingSourc
+            if (MappingLookup.fromMapping(incomingMapping).exceedsLimit(maxFieldsToAddDuringMerge, 0)) {
+                newMapping = Mapping.EMPTY.merge(incomingMapping, reason, maxFieldsToAddDuringMerge);
+            } else {
+                newMapping = incomingMapping;
```
Yes, we could do that. The else-branch is basically an optimization that prevents a merge process if it isn't necessary. But I'm equally happy to simplify the code.

javanna replied:
ok, I was maybe making the wrong assumption that that merge would mostly be a no-op.

javanna replied:
Ok, refreshing my memory: the current mapper is null when the index has no mappings, and that conditional aims to cover the scenario where the first document getting indexed goes already beyond the limit or not. I think I would remove this optimization and always call EMPTY.merge. Perhaps we could think about overriding `merge` for that specialized empty mapping and not doing any work, but it's final and I am not sure we want to make it non-final.

By the way, how are we going to make sure that this is not exercised by static mappings as well? It currently is, isn't it? That reminds me we have so many different merge methods and that makes things very difficult to follow. Which merge is used where? I will need to review that bit once we get there.

felixbarny replied:
> I think I would remove this optimization and always call EMPTY.merge.

I've implemented that and it actually revealed a bug that was caught by a lot of test cases. Good call to always call the same code path so that there's not a code path that's rarely executed. I've fixed the bug in 48e2f33.

> Perhaps we could think about overriding merge for that specialized empty mapping and not doing any work, but it's final and I am not sure we want to make it non-final.

The merge is short-circuited in `MergeResult#addNewObjectMapper` so it won't recurse all the way down. Seems like that's enough of an optimization.

javanna replied:
Tell me more about the bug? I was going to ask you why you went from Mapping.merge to root.withoutMapper.merge. It's subtle!

felixbarny replied:
If `enabled` is set to `false` in the root object mapper, the following exception was thrown:

https://github.com/elastic/elasticsearch/blob/9e074b8e7521129504024cc2b7b446c3485611da/server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java#L499

That's because `Mapper.EMPTY` has an implicit value for `enabled` of `true`. Thus, you can't merge `Mapper.EMPTY` with mappings that have set `enabled` to `false`.

felixbarny committed:
Merge remote-tracking branch 'origin/main' into mapping-merge-limit-fields

felixbarny commented on server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:
```diff
@@ -558,12 +558,25 @@ public Mapping parseMapping(String mappingType, Map<String, Object> mappingSourc
     }
 
-    public static Mapping mergeMappings(DocumentMapper currentMapper, Mapping incomingMapping, MergeReason reason) {
+    public Mapping mergeMappings(Mapping incomingMapping, MergeReason reason) {
```
The method has exactly 4 callers and all of them just pass in `maapperService.documentMapper()`, so having this as a static method doesn't seem to make much sense, IMHO. There's still a package-private static method that's used for tests.

> Potentially, this static method could be moved away from MapperService etc. but we never got to making that change.

We can always re-introduce that once it's needed. As there's still package-private method, that should be straightforward to do when needed.

felixbarny replied:
To put this another way, I haven't really removed the static method but just created an instance method for convenience, that delegates to the static method, using `this.mapper` as for the `currentMapper` parameter.

javanna replied:
I know, it's subtle, and while the current state of things needs more work, I prefer the public method to be static. You can view that as an effort to keep the surface of MapperService smaller. There are way too many merge methods and making this public mergeMappings was a first step trying to move merging away from requiring an instance of MapperService (in fact in the master we create a mapper service only to do the merging, which should not be required).

felixbarny replied:
gotcha, I'll revert

felixbarny replied:
Reverted in d3c962b

felixbarny committed:
Update parameter names to mapperMergeContext

felixbarny committed:
Add Javadoc to MapperMergeContext and make it final

felixbarny committed:
Make MapperService#mergeMappings static again

felixbarny committed:
Merge remote-tracking branch 'origin/main' into mapping-merge-limit-fields

felixbarny committed:
Make field adding methods in MapperMergeContext package-private

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java:
```diff
@@ -281,6 +281,10 @@ void checkFieldLimit(long limit, int additionalFieldsToAdd) {
         }
     }
 
+    boolean exceedsLimit(long limit, int additionalFieldsToAdd) {
```
if we don't use it and rely on merge alone, even for the first document being indexed, let's revert changes in this class?

javanna replied:
got it. Leave it then.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:
```diff
@@ -403,6 +403,10 @@ public ObjectMapper.Builder newBuilder(IndexVersion indexVersionCreated) {
         return builder;
     }
 
+    public ObjectMapper withoutMappers() {
```
Does this here and in the subclasses need to be public? Can you also use the `@Override` annotation in the subclasses that override it?

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:
```diff
@@ -533,15 +537,22 @@ private static Map<String, Mapper> buildMergedMappers(
-            Map<String, Mapper> mergedMappers = null;
+            Map<String, Mapper> mergedMappers = new HashMap<>(existing.mappers);
             for (Mapper mergeWithMapper : mergeWith) {
-                Mapper mergeIntoMapper = (mergedMappers == null ? existing.mappers : mergedMappers).get(mergeWithMapper.simpleName());
```
I believe that we are attempting not to copy the map into a new one here. See #83776 . I am not sure if this is necessary, but I would leave it as-is unless if possible. oh well actually down below we do the same copy. That's odd. If you were cleaning things up, could we discuss those changes separately unless they are necessary to make your change work?

felixbarny replied:
I've extracted some of the mechanical changes in this PR: https://github.com/elastic/elasticsearch/pull/104578

felixbarny replied:
Looks much cleaner now that that PR has been merged.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:
```diff
@@ -533,15 +537,22 @@ private static Map<String, Mapper> buildMergedMappers(
+                        ObjectMapper withoutMappers = om.withoutMappers();
+                        boolean added = objectMergeContext.addFieldIfPossible(mergedMappers, withoutMappers);
+                        if (added) {
+                            mergedMappers.put(om.simpleName(), withoutMappers.merge(mergeWithMapper, reason, objectMergeContext));
```
are we not putting twice now? Once within addFieldIfPossible and once here ? The value is different but the key is the same?

felixbarny replied:
Hopefully, `MergeResult#addNewObjectMapper` also makes this a little clearer now. But It still essentially puts the same key twice. We could prevent that by exposing a method that decrements the field budget but doesn't add the Mapper to the `mergedMappers` map. However, I'd prefer not exposing such a method so that the process of checking the budget, decrementing it, and executing a consumer that puts the mapper to the `mergedMappers` map is encapsulated and callers can't forget a step.

If you feel strongly that this is not a good idea or not readable, we could also expose `MapperMergeContext.NewFieldsBudget#decrementIfPossible` and make the caller responsible for putting the mapper to the `mergedMappers` map.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:
```diff
@@ -553,22 +564,13 @@ private static Map<String, Mapper> buildMergedMappers(
+                        mergedMappers.put(mergeWithMapper.simpleName(), mergeWithMapper);
                     } else {
-                        merged = mergeIntoMapper.merge(mergeWithMapper, objectMergeContext);
+                        mergedMappers.put(mergeWithMapper.simpleName(), mergeIntoMapper.merge(mergeWithMapper, objectMergeContext));
```
Is it necessary to move where the put happens (within each if branch compared to at the end)? I am trying to figure out if that is just noise in the diff. If possible, I'd like to keep things as they were as much as possible.

felixbarny replied:
I've tried again, but the recursive nature of this, combined with the fact that we want to limit fields here makes this challenging.

If we only add the merged mappers to the map at the end of the method, and if we decrement the number of fields at the end, we would first consume the budget for the child mappers so that when the recursion ends and we're putting the mappers into the map, the child mappers may have consumed all of the budget so that we can't add the parent objects in.

The intention of the `MapperMergeContext.addFieldIfPossible` was to keep the act of checking the limit, adding the mapper to the map, and decrementing the budget as close to one another as possible. Otherwise, it can be easy to miss a step.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:
```diff
@@ -533,15 +537,22 @@ private static Map<String, Mapper> buildMergedMappers(
                 if (mergeIntoMapper == null) {
-                    merged = mergeWithMapper;
+                    if (mergeWithMapper instanceof ObjectMapper om) {
+                        ObjectMapper withoutMappers = om.withoutMappers();
```
Could you help me understand the need for `withoutMappers` here please as well as the two branches of this conditional?

`mergeIntoMapper` is null, hence we are adding a new field, which I suspect is why we don't go through merge, we need to decrement the budget by the size of the current mapper don't we? But isn't the budget always one using `withoutMappers`? I see that we later use merge, which should take the limit into account as well, but then I don't understand the need for the initial `addFieldIfPossible` call before `merge`.

felixbarny replied:
I've added some comments to hopefully make it a little clearer.

felixbarny replied:
I think the new `MergeResult#addNewObjectMapper` method makes this a bit clearer, too.

javanna reviewed (changes requested):
I completed my in-depth review round and left a bunch of comments and questions.

felixbarny committed:
Apply feedback from review

felixbarny committed:
Use Consumer in addFieldIfPossible

Avoid passing in Mappers map into addFieldIfPossible,
instead rely on the consumer to do that
Added some comments to the merge process

felixbarny committed:
Extract ObjectMapper#addNewObjectMapper

As an additional optimization, try to add the full object first.
Only merge sub-mappers one by one in case there wasn't enough budget for the full object mappper.

felixbarny committed:
Make `FieldMapper.MultiFields.Builder.add(FieldMapper)` private

felixbarny committed:
Merge remote-tracking branch 'origin/main' into mapping-merge-limit-fields

felixbarny committed:
Fix merge with empty mapping

felixbarny committed:
Attempt to fix mapper serialization assertion

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/Mapping.java:
```diff
@@ -135,8 +135,9 @@ public SourceLoader.SyntheticFieldLoader syntheticFieldLoader() {
      */
-    Mapping merge(Mapping mergeWith, MergeReason reason) {
-        RootObjectMapper mergedRoot = root.merge(mergeWith.root, reason, MapperMergeContext.root(isSourceSynthetic(), false));
+    Mapping merge(Mapping mergeWith, MergeReason reason, long maxFieldsToAddDuringMerge) {
```
++ to newFieldsBudget I think it's clear and concise.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/MapperMergeContext.java:
```diff
@@ -16,25 +19,31 @@
 public final class MapperMergeContext {
 
     private final MapperBuilderContext mapperBuilderContext;
+    private final AtomicLong remainingFieldsUntilLimit;
```
AtomicLong has also some overhead to to it being thread-safe, and can be misleading to see it used in non concurrent scenario. After all, it is a long counter with some sugar on top but no need for atomicity in this code. Could we use a long counter?

felixbarny replied:
It needs to be a mutable reference to a `long` but I've encapsulated it in a `NewFieldsBudget` interface that has different implementations for the unlimited and limited cases which should also address some other concerns that you had.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/MapperMergeContext.java:
```diff
@@ -43,10 +52,57 @@ private MapperMergeContext(MapperBuilderContext mapperBuilderContext) {
+        return false;
+    }
+
+    void addFieldIfPossible(Mapper mapper, Runnable addField) {
```
The reason why I suggested the boolean approach is that we have now two methods that do the same, but not quite. I think it would be nice to consolidate that to a single method if possible. After all, if we return a boolean in one method, it is already responsibility of the caller to act on the return value, but the budget check is done within the method, where it belongs.

felixbarny replied:
> The reason why I suggested the boolean approach is that we have now two methods that do the same, but not quite. I think it would be nice to consolidate that to a single method if possible.

The new `NewFieldsBudget` interface now has a `decrementIfPossible(long fieldSize)` method that does what you proposed. I've still kept the `addFieldIfPossible(M mapper, Consumer<M> addField)` method as the only public method in `MapperMergeContext` that internally calls `decrementIfPossible(long fieldSize)` and also calls the consumer if there was enough budget. I do like the encapsulation of checking the budget, decrementing the budget, and adding the Mapper to the map in a single call.

>  After all, if we return a boolean in one method, it is already responsibility of the caller to act on the return value, but the budget check is done within the method, where it belongs.

In most cases, the caller doesn't need to know whether the Consumer has been called. But in some instances, it's important. But I think it shouldn't be the responsibility of the caller to add the mapper. It seems more brittle to me.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/MapperMergeContext.java:
```diff
@@ -43,10 +52,57 @@ private MapperMergeContext(MapperBuilderContext mapperBuilderContext) {
+        } else if (canAddField(1)) {
+            remainingFieldsUntilLimit.decrementAndGet();
+            runtimeFields.put(runtimeField.name(), runtimeField);
+        }
```
yea I guess I forgot the canAddField call, but the point is that we can potentially call put in one place. It's not super important, I just thought it would make the code a tiny bit clearer, maybe it does not make a difference.

felixbarny replied:
Not sure if that's feasible as we're not calling `put` in all branches. In the `else` case (trying to add a new runtime field when the field limit is reached), we don't call `put`. However, this method is now removed anyway, resolving.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:
```diff
@@ -559,11 +559,24 @@ public Mapping parseMapping(String mappingType, Map<String, Object> mappingSourc
+        DocumentMapper currentMapper,
+        Mapping incomingMapping,
+        MergeReason reason,
+        long maxFieldsToAddDuringMerge
```
Additional question: how about the scenario where dynamic beyond limit is not used? Would we be counting for nothing? This seems like a scenario that would affect dynamic mappings, hence indexing of every document, which we should try and avoid.

felixbarny replied:
There are now two implementations of the `NewFieldsBudget` class. With `NewFieldsBudget.unlimited()`, you get an instance that doesn't maintain a counter.

felixbarny committed:
Apply spotless suggestions

felixbarny committed:
Add NewFieldsBudget class

felixbarny committed:
Merge remote-tracking branch 'origin/main' into mapping-merge-limit-fields

felixbarny commented on server/src/main/java/org/elasticsearch/index/mapper/MapperMergeContext.java:
```diff
@@ -43,10 +52,57 @@ private MapperMergeContext(MapperBuilderContext mapperBuilderContext) {
+
+    void removeRuntimeField(Map<String, RuntimeField> runtimeFields, String name) {
+        if (runtimeFields.containsKey(name)) {
+            runtimeFields.remove(name);
```
I'd prefer to keep that out of the scope of this PR.

felixbarny committed:
Relax assertion in RootObjectMapper#merge

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:
```diff
@@ -543,7 +547,14 @@ private static Map<String, Mapper> buildMergedMappers(
+                    } else {
+                        objectMergeContext.addFieldIfPossible(
+                            mergeWithMapper,
+                            m -> mergedMappers.put(mergeWithMapper.simpleName(), mergeWithMapper)
```
if you keep the consumer, it may make sense to use `m` instead of `mergeWithMapper`, although they are in practice the same. Maybe using a consumer is flawed :)

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:
```diff
@@ -543,7 +547,14 @@ private static Map<String, Mapper> buildMergedMappers(
                 if (mergeIntoMapper == null) {
-                    mergedMappers.put(mergeWithMapper.simpleName(), mergeWithMapper);
+                    if (mergeWithMapper instanceof ObjectMapper om) {
+                        addNewObjectMapper(reason, objectMergeContext, mergedMappers, om);
```
I think I'd slightly rewrite this to do the common `addFieldIfPossible` where the put is currently done , and only if that one fails, do the conditional fallback to the object mapper specific logic.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:
```diff
@@ -565,6 +576,30 @@ private static Map<String, Mapper> buildMergedMappers(
+            }
+            // there's not enough capacity for the whole object mapper,
+            // so we're just trying to add the shallow object, without it's sub-fields
+            ObjectMapper shallowObjectMapper = objectMapper.withoutMappers();
```
is there a test that fails if this fallback is entirely removed? Do we verify that it does what it's supposed to do? I am thinking that despite the index setting is not yet wired, we can add unit tests to ObjectMapperMergeTests with an real budget and leverage this codepath?

felixbarny replied:
Yes, I've already added `org.elasticsearch.index.mapper.ObjectMapperMergeTests#testMergeWithLimitObjectField` to test for this case.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/MapperMergeContext.java:
```diff
@@ -43,10 +49,113 @@ private MapperMergeContext(MapperBuilderContext mapperBuilderContext) {
         return mapperBuilderContext;
     }
+
+    void removeRuntimeField(Map<String, RuntimeField> runtimeFields, String name) {
```
Now that I think of this more: removing a runtime field can only be done via static declaration of mappings, hence I don't think it should be accounted for here? you need to set the field to null in a put mapping call... removing this would spare us the need for the increment method in the budget.

I was also thinking: once we reach the limit, we are never going to go backwards, meaning the budget can effectively only ever be decremented (if we agree that remove should go away)? In that case, we could consider swapping the budget to a no-op that always returns false, instead of checking of nothing: we already know we reached the limit? Does this make sense?

javanna replied:
Nevermind, what I suggested is probably overkill.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:
```diff
@@ -565,6 +576,30 @@ private static Map<String, Mapper> buildMergedMappers(
             return Map.copyOf(mergedMappers);
         }
+
+        private static void addNewObjectMapper(
```
Thanks, this method and extracting #104578 are great ideas and do help following what is happening in this PR.

javanna reviewed (commented):
I did another round. Thanks for all the progress you made. Don't get discouraged by the amount of comments, these are all details around refining, polishing and adding some tests. We are very close!

Could you also add docs to clarify design choices: why the field limit is checked as part of merge. I could see future readers getting confused about that. And describe how it works, meaning it is already applied in the data node, and if certain fields are allowed there, they are going to be allowed when the master does the merge of the mappings, but there are scenarios where data nodes allowed addition of fields that the master will reject. Am I getting this right? Could you expand on the mechanics of this, perhaps in the javadocs of MapperMergeContext?

felixbarny commented on server/src/main/java/org/elasticsearch/index/mapper/ParsedDocument.java:
```diff
@@ -160,7 +161,7 @@ public void addDynamicMappingsUpdate(Mapping update) {
             dynamicMappingsUpdate = update;
         } else {
-            dynamicMappingsUpdate = dynamicMappingsUpdate.merge(update, MergeReason.MAPPING_UPDATE);
+            dynamicMappingsUpdate = dynamicMappingsUpdate.merge(update, MergeReason.MAPPING_UPDATE, NewFieldsBudget.unlimited());
```
No, we'll apply the field limit in `MapperService#merge`, depending on whether the merge reason resembles an auto-update (dynamic mappings). See also https://github.com/elastic/elasticsearch/pull/96235/files#diff-32c1e36544e0c0ef775f9e26576ef7df0504cc05466e513ba7b143947c234784R590.

The only instance when `addDynamicMappingsUpdate` is called is during `prepareIndex` to ensure that there's an empty mapping:

https://github.com/elastic/elasticsearch/blob/78c6a7c4bb177d9e63f7c947a09e75cda6a616de/server/src/main/java/org/elasticsearch/index/shard/IndexShard.java#L1008-L1019

That's why this will remain to use no limit.

Ideally, I'd like to remove the method tbh or don't merge or override the mapping within `addDynamicMappingsUpdate`.

felixbarny commented on server/src/main/java/org/elasticsearch/index/mapper/MapperMergeContext.java:
```diff
@@ -43,10 +49,113 @@ private MapperMergeContext(MapperBuilderContext mapperBuilderContext) {
+     * @return a new {@link MapperMergeContext}, wrapping the provided {@link MapperBuilderContext}
+     */
+    public MapperMergeContext createChildContext(MapperBuilderContext childContext) {
+        return new MapperMergeContext(childContext, remainingFieldsBudget);
```
done

felixbarny replied:
> ambiguity between creating a child mapper merge context and a child mapper builder context.

A child mapper merge context includes creating a child mapper builder context, as well as passing the field budget down to the child.

felixbarny commented on server/src/main/java/org/elasticsearch/index/mapper/MapperMergeContext.java:
```diff
@@ -43,10 +49,113 @@ private MapperMergeContext(MapperBuilderContext mapperBuilderContext) {
+    }
+
+    /**
+     * Keeps track of now many new fields can be added during mapper merge
```
done

felixbarny commented on server/src/main/java/org/elasticsearch/index/mapper/MapperMergeContext.java:
```diff
@@ -43,10 +49,113 @@ private MapperMergeContext(MapperBuilderContext mapperBuilderContext) {
+    /**
+     * Keeps track of now many new fields can be added during mapper merge
+     */
+    public interface NewFieldsBudget {
```
I've made it a private interface now. If we wanted to keep it public and pass it around instead of a long, we could also make it a sealed interface that only permits the `Unlimited` and `Limited` implementations.

felixbarny committed:
Apply feedback from review

felixbarny commented on server/src/main/java/org/elasticsearch/index/mapper/Mapping.java:
```diff
@@ -133,10 +134,12 @@ public SourceLoader.SyntheticFieldLoader syntheticFieldLoader() {
      */
-    Mapping merge(Mapping mergeWith, MergeReason reason) {
-        RootObjectMapper mergedRoot = root.merge(mergeWith.root, reason, MapperMergeContext.root(isSourceSynthetic(), false));
+    Mapping merge(Mapping mergeWith, MergeReason reason, NewFieldsBudget newFieldsBudget) {
```
I've made `NewFieldsBudget` private and reverted to use `long`s in method arguments.

felixbarny commented:
> Could you also add docs to clarify design choices: why the field limit is checked as part of merge. I could see future readers getting confused about that. And describe how it works, meaning it is already applied in the data node, and if certain fields are allowed there, they are going to be allowed when the master does the merge of the mappings, but there are scenarios where data nodes allowed addition of fields that the master will reject. Am I getting this right? Could you expand on the mechanics of this, perhaps in the javadocs of MapperMergeContext?

I've documented most of that in the other PR. See also https://github.com/elastic/elasticsearch/pull/96235/files#diff-32c1e36544e0c0ef775f9e26576ef7df0504cc05466e513ba7b143947c234784R590

felixbarny committed:
Fix checkstyle issue

felixbarny commented on server/src/main/java/org/elasticsearch/index/mapper/MapperMergeContext.java:
```diff
@@ -43,10 +49,113 @@ private MapperMergeContext(MapperBuilderContext mapperBuilderContext) {
+        if (remainingFieldsBudget.decrementIfPossible(1)) {
+            addField.accept(runtimeField);
+        }
+    }
```
I've removed the Consumer methods in favor of a single `decrementIfPossible` method.

javanna replied:
thanks

felixbarny committed:
Rename method to MapperMergeContext#decrementFieldBudgetIfPossible

felixbarny committed:
Move put method out of the if/else again

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/Mapper.java:
```diff
@@ -135,4 +135,18 @@ public static FieldType freezeAndDeduplicateFieldType(FieldType fieldType) {
+     * <p>
+     * Needs to be in sync with {@link MappingLookup#getTotalFieldsCount()}.
+     */
+    public int mapperSize() {
```
It's fine. leave it and rename it later.

javanna commented on server/src/test/java/org/elasticsearch/index/mapper/ObjectMapperMergeTests.java:
```diff
@@ -198,6 +204,66 @@ public void testMergedFieldNamesMultiFieldsWithinSubobjectsFalse() {
+        ObjectMapper mergedAdd1 = mergeInto.merge(mergeWith, MapperMergeContext.root(false, false, 1));
+        assertEquals(2, mergedAdd0.mapperSize());
+        assertEquals(3, mergedAdd1.mapperSize());
+    }
```
Should we expand these tests to cover for the following scenarios? 

- add new runtime fields - budget is taken into account
- add existing runtime fields - budget does not matter
- perhaps check that removing is not affected by the budget in any way, or maybe this is difficult to test, not sure
- adding the same object but with new sub-fields
- adding the same object with exactly the same sub-fields as already existing
- different combinations of subobjects and enabled perhaps

felixbarny replied:
makes sense, I'll look into it.

felixbarny replied:
I've added additional test cases.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:
```diff
@@ -557,14 +566,28 @@ private static Map<String, Mapper> buildMergedMappers(
+                // now trying to add the sub-fields one by one via a merge, until we hit the limit
+                return shallowObjectMapper.merge(objectMapper, reason, context);
+            }
+            return null;
```
do we test for this specific case where there is no budget to even add the container?

felixbarny replied:
I've added a case in `org.elasticsearch.index.mapper.ObjectMapperMergeTests#testMergeWithLimitObjectField` now.

javanna reviewed (approved):
I did a last round, this look really good. My final comments are mainly on strengthening the tests further and a tiny bit more docs. Thanks for all the work. No need for another review round, merge when ready.

felixbarny commented on server/src/main/java/org/elasticsearch/index/mapper/RootObjectMapper.java:
```diff
@@ -242,12 +258,16 @@ public RootObjectMapper merge(Mapper mergeWith, MergeReason reason, MapperMergeC
         }
         final Map<String, RuntimeField> runtimeFields = new HashMap<>(this.runtimeFields);
-        assert this.runtimeFields != mergeWithObject.runtimeFields;
+        assert this.runtimeFields != mergeWithObject.runtimeFields || this.runtimeFields == Map.<String, RuntimeField>of();
```
Both `this.runtimeFields` and `mergeWithObject.runtimeFields` can have the same reference - to `Map.of()` (immutable empty map). In this case we shouldn't throw an assertion error.

This happens in the context of `org.elasticsearch.index.mapper.Mapping#withFieldsBudget`. We're calling `org.elasticsearch.index.mapper.RootObjectMapper#withoutMappers` which provides `Map.of()` for the `runtimeFields` argument. Even if we changed that and provided a `new HashMap<>()`, we'd run into a similar situation because in `org.elasticsearch.index.mapper.RootObjectMapper#merge`, we're doing `Map.copyOf(runtimeFields)`, which returns the equivalent of `Map.of()` in case you provide an empty map.

Maybe we should just drop the assertion entirely?

javanna replied:
Thanks for the context. The assert was added in #65378 , back when we were still cloning objects mappers. Clone has been removed with #101485 , hence I think the assert can be removed too.

felixbarny replied:
done

felixbarny commented on server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:
```diff
@@ -403,6 +403,10 @@ public ObjectMapper.Builder newBuilder(IndexVersion indexVersionCreated) {
         return builder;
     }
 
+    ObjectMapper withoutMappers() {
```
done

felixbarny commented on server/src/main/java/org/elasticsearch/index/mapper/NestedObjectMapper.java:
```diff
@@ -186,6 +186,21 @@ public ObjectMapper.Builder newBuilder(IndexVersion indexVersionCreated) {
     }
 
+    @Override
+    NestedObjectMapper withoutMappers() {
```
It requires a lot of boilerplate and setup to make sure all properties are populated. But I can give it a try.

felixbarny replied:
done

felixbarny commented on server/src/test/java/org/elasticsearch/index/mapper/ObjectMapperMergeTests.java:
```diff
@@ -198,6 +204,66 @@ public void testMergedFieldNamesMultiFieldsWithinSubobjectsFalse() {
+        ObjectMapper mergedAdd3 = root.merge(mergeWith, MapperMergeContext.root(false, false, 3));
+        assertEquals(0, root.mapperSize());
+        assertEquals(1, mergedAdd1.mapperSize());
+        assertEquals(2, mergedAdd2.mapperSize());
```
Good catch. I've updated the test name.

felixbarny commented on server/src/test/java/org/elasticsearch/index/mapper/ObjectMapperMergeTests.java:
```diff
@@ -198,6 +204,66 @@ public void testMergedFieldNamesMultiFieldsWithinSubobjectsFalse() {
+        assertEquals(3, rootObjectMapper.mapperSize());
+        assertEquals(4, mergeWith.mapperSize());
+        assertEquals(3, mergedAdd0.mapperSize());
+        assertEquals(4, mergedAdd1.mapperSize());
```
It's a bit tricky as it's not deterministic which fields are added. But the next test does some of that already, so I don't think there's a need to do that in this test, too.

felixbarny committed:
Enhance javadoc

felixbarny committed:
Enhance testMergeWithLimitTruncatedObjectField

felixbarny committed:
Remove assertion

felixbarny committed:
Add more cases to ObjectMapperMergeTests

felixbarny committed:
Add test cases for ObjectMapper#withoutMappers

felixbarny committed:
Merge remote-tracking branch 'origin/main' into mapping-merge-limit-fields
